docs/poly-to-normal-variant/index.md

```diff
@@ -0,0 +1,345 @@
+tags don't need to be fully specified:
+
+<<< Types.re#delete-refinement{4-11}
+
+The variant tags inside `DiscountError` originate from our `Discount` module, so
+they shouldn't be needlessly constrained by the `discount` type defined inside
+the `Promo` module.
+
+Change the `discount` type to this:
+
+<<< Types.re#type-variable
+
+Now `discount` is a *type constructor* that takes a *type variable* named `'a`.
+A type constructor is not a fixed type---you can think of it as a function that
+takes a type and outputs a new type.
+
+Using a type variable does not sacrifice type safety, if you hover over the
+`discount` variable, you see that its type is now:
+
+```reason
+discount([> `MissingSandwichTypes
+          | `NeedMegaBurger
+          | `NeedOneBurger
+          | `NeedTwoBurgers ])
+```
+
+This is essentially the same type as before (other than the `>` symbol,
+explained in the next section), just written differently. By looking at the
+usage of the `discount` variable, OCaml can infer how to fill in the type
+variable and produce the fixed type shown above.
+
+The advantage of using a type variable for the definition of the
+`Promo.discount` type is that when you add, rename, or delete variant tags in
+`Discount`, you won't have to make corresponding edits to `Promo.discount`[^1].
+
+## `>` = "allow more than"
+
+In the type expression above, we once again see `>`, so let's see what it means.
+In polymorphic variant type expressions, it means "allow more than". In this
+case, it means that tags other than the four that are listed are allowed. For
+example, this type would be allowed:
+
+```reason{5-6}
+discount([| `MissingSandwichTypes
+          | `NeedMegaBurger
+          | `NeedOneBurger
+          | `NeedTwoBurgers
+          | `HoneyButter
+          | `KewpieMayo ])
+```
+
+When defining your own types, you will most often used *fixed* polymormorphic
+variants, i.e. those that don't have `>` in their type expressions. But it is
+still useful to know what `>` does, since it appears when the compiler
+infers the type of a variable or function that uses polymorphic variants.
+
+::: tip
+
+Fixed polymorphic variants and normal variants are roughly equivalent and can be
+used interchangeably.
+
+:::
+
+## Implicit type variable
+
+Let's come back to the question of why the original attempt at a variant type
+definition was syntactically invalid:
+
+<<< Types.re#bad-discount-type
+
+The reason is that there's an implicit type variable around the `>`. The
+above type expression is equivalent to:
+
+<<< Types.re#explicit-type-var{14}
+
+Now the error message makes a bit more sense:
+
+<<< type-error.txt{7}
+
+The type variable exists, but it's pointless unless it appears as an argument of
+the `discount` type constructor. Once it's added, it compiles:
+
+<<< Types.re#add-type-arg{1}
+
+This is somewhat like accidentally using a variable in a function but forgetting
```

Comment:
Something that helps me wrap my head around these arguments is that they will be used by the compiler to "carry" the information about which extra tags are being added to the open variant, on each usage.